### ISSUE TITLE
Add search option

### ISSUE THREAD

**Issue by boesiii** [tomchadwin/qgis2web#294]
A user asked if I could create some type of query option so they could find a point with an ID equal to something and the map would zoom to a particular feature with matching attribute.

I have something working for leaflet based on
https://github.com/stefanocudini/leaflet-search

Would this be something worth incorporating into Q2W?

**Comment by tomchadwin** [comment]
I wanted to include Stefano's plugin to allow search on attributes. From memory, there was an issue with the feature highlight style applied to search results never resetting. Do I take it that it a works for you?

**Comment by boesiii** [comment]
When a query is performed and it finds a feature it adds a pink circle around the found feature.  There is no way to clear the pink circle unless you perform a different query and therefore the pink circle moves.  Is that what you area talking about?

**Comment by boesiii** [comment]
Assuming I could get this to function correctly.  How would you expect this to work in the q2w GUI?  Would it be in the layer list?  Would the query tool have it's own area?

**Comment by tomchadwin** [comment]
Yes, that's the style reset issue I had. Of that were fixed, we should definitely look at incorporating it. 

As for the GUI, I'm not immediately sure without more thought as to how it should work. A new setting in the bottom-left pane  with a dropdown of the layers, and a second dropdown which auto-populates with the selected layer's fields would be one way to do it. That would limit the search to just one layer, without a lot of complexity. 

Any thoughts?

**Comment by tomchadwin** [comment]
Not sure how easy this would be, but how about adding another column to each layer's `Popup fields` section in the GUI, and putting a checkbox, with the heading "Search", or a search icon? That would allow any number of fields to be searchable.

That said, does the Leaflet plugin support searching on multiple fields. Plus we have to have an OL3 solution as well.

**Comment by boesiii** [comment]
With my hack programming skills both will be difficult!

This shows how to search multiple layers
http://labs.easyblog.it/maps/leaflet-search/examples/multiple-layers.html

**Comment by boesiii** [comment]
I figured out how to reset the style where it draws a circle around the found feature.  The plugin has an option to hide and the query style when search collapses.

**Comment by tomchadwin** [comment]
Ah, that sounds promising. Are you able to add the JS into a qgis2web map, so I can look at how we might implement it?

**Comment by boesiii** [comment]
Here is a sample
[qgis2web_query.zip](https://github.com/tomchadwin/qgis2web/files/462129/qgis2web_query.zip)

**Comment by tomchadwin** [comment]
Thanks! Questions:

1. That seems to reset the style correctly - have you hacked the search lib itself to do that as you described? In other words, have you included a customized version of @stefanocudini's plugin?
2. How does it know which field to search? It's only searching `NAME`

**Comment by tomchadwin** [comment]
Ah, it's OK for 1. above. You changed the default in the library, but you can just add it as an option when you instantiate it in `index.html`:

```javascript
map.addControl( new L.Control.Search({layer: json_airports0JSON, hideMarkerOnCollapse: true}) );
```

What about 2, above?

**Comment by boesiii** [comment]
I modified the original plugin js code.  Look at line 58

`propertyName: 'NAME'`

Which could probably be able to be included similar to your comment above

`map.addControl( new L.Control.Search({layer: json_airports0JSON, hideMarkerOnCollapse: true, propertyName: 'NAME'}) );`

**Comment by tomchadwin** [comment]
Got it. Thanks. So just need to crack multiple fields/layers, and we can implement this for Leaflet. 

@ThomasG77, what's the approved way to implement a text search on attribute fields in OL3? I think there's an example on the site. Is that the best place to start? A plugin would be better.

**Comment by ThomasG77** [comment]
For OL3, I'm not aware of a definitive "approved way".
There are "search like" components in particular for geocoding but I'm not aware there are some for going through a vector source from a layer. A beginning could be below code 

```
function match_col_value_for_layer(options) {
  var source = options.source;
  var col = options.col
  var text = options.text;
      text = text.replace(/[.*+?^${}()|[\]\\]/g, '');
  var features = [];
  var I = options.initial ? '^' : '';  //search only initial text
  var icase = !options.casesensitive ? 'i' : undefined;
  var regSearch = new RegExp(I + text, icase);

  source.forEachFeature(function(feature) {
    if (regSearch.test(feature.get(options.col))) {
      features.push(feature);
    }
  });
  return features;
}

// Go on http://openlayers.org/en/latest/examples/image-vector-layer.html

var filteredFeats = match_col_value_for_layer({
  source: map.getLayers().item(1).getSource().getSource(),
  col: 'name',
  text: 'AFg'
});

// TODO
// Reuse the result to create a list for dynamic autocomplete and do not forget to manage how to zoom on feature and/or highlight them.
```
You should also think about the wanted function: a search (the plugins you mentioned) or a filtering e.g an input where you limit results in a list like on the left of http://bmcbride.github.io/bootleaf/

**Comment by tomchadwin** [comment]
@ThomasG77 Thanks! Despite the title of this issue, it's search I'm after - you too, @boesiii? We already have geocoding. Will have a look at your snippet and see if I can do anything with it.

**Comment by tomchadwin** [comment]
Back to Leaflet for the moment. If I understand the plugin correctly, it can search multiple layers, if they are in a layer group, but it can only search a single property. That means it will only work across multiple layers if the property name is the same in those layers. Is that right, @stefanocudini?

The good news is the the qgis2web Leaflet export already places all layers in a layer group called `feature_group`, so we could just pass that:

```javascript
map.addControl( new L.Control.Search({layer: feature_group,
                                              hideMarkerOnCollapse: true,
                                              propertyName: 'NAME'}) );
```

However, i think for the moment, if I am right about the inability to search multiple properties, the best way to implement this in the GUI is going to be a single drop-down in the bottom-left pane:

Search field:
none
layer1: field1
layer1: field2
layer2: field1
layer2: field2

Does that make sense?

**Comment by ThomasG77** [comment]
I've committed a plugin. At the moment, it takes only one layer and one column.
I'm not sure a classic loop through each layer and each column is efficient.

Something like indexation should be needed for performances.
![demo-ol-search-layer](https://cloud.githubusercontent.com/assets/642120/18384890/8db115ba-768d-11e6-8b63-a9ce3df9636f.gif)

**Comment by tomchadwin** [comment]
@ThomasG77 Fabulous! Single-layer, single column is fine to begin with - see my previous description of how I think the GUI will work.

Where is the plugin?

**Comment by boesiii** [comment]
@tomchadwin Yes, should have titled it search instead of query.

**Comment by tomchadwin** [comment]
Leaflet search added to master. Please test.

**Comment by boesiii** [comment]
I get an error in the browser.

`L.Control.Search is not a constructor  Line 126`

I don't think feature_group can be the value for layer.

**Comment by boesiii** [comment]
I am not seeing the search icon on the web page

**Comment by tomchadwin** [comment]
Ah, I wonder. Without losing any of your own template customizations (take a backup), can you copy the contents of `.qgis2\python\plugins\qgis2web\templates` to overwrite `.qgis2\qgis2web\templates`? Does that fix it?

**Comment by boesiii** [comment]
Yes that fixed it.  Why?

What is `.qgis2\qgis2web\` used for?

**Comment by boesiii** [comment]
Yes #2.

Instead of overwriting the files what about renaming each original file by adding a time and date stamp.

`full-screen 2016-09-09 1241.html`

**Comment by tomchadwin** [comment]
Yes, but only if the old and new template files differ:

1. If templates do not exist, copy them
2. If identical templates exist, do not copy them
3. If different templates exist, rename them and copy the new ones up

3 could be enhanced as follows:

3. If different templates exist, rename them and copy the new ones up, unless they match the previous template versions, in which case just overwrite them

OK, I'll look at that next week.

**Comment by ThomasG77** [comment]
@tomchadwin At least published https://github.com/webgeodatavore/ol3-search-layer.
I plan to change quite a lot of things (more a first messy version than a release I could be proud).
Feedback welcome (missing features or options, buggy behaviour,...)

**Comment by tomchadwin** [comment]
Splendid. I'll look into it next week. Thanks!

**Comment by tomchadwin** [comment]
OK, I've committed @ThomasG77's OL3 attribute search plugin to master. However, I can't get it to work. The search icon and field appear, but nothing happens when you try to search. No Python console errors, no JS console errors, no network errors. I can't figure out why it's not working. Any ideas, anyone?

In an earlier version of the the OL3 search plugin README, it said that it only returned a list of features - that it was up to you to implement a function to use that list. However, I can't find that wording in the README any more.

In the example on linked from the repo, I don't see any additional functions called on search completion - ie highlight and zoom to feature. Am I missing something from the example?

PS  This is also failing Travis tests, but it seems/feels as though this is intermittent. Every test fails, but the plugin works fine locally. If anyone can explain the Travis issue, I'd also be extremely grateful.

**Comment by tomchadwin** [comment]
Tests fixed. Still not working.

**Comment by tomchadwin** [comment]
Polyfills added, which I thought was the issue. But it's still not working (by the way, not updated the test control files with the polyfills, since I don't know if they are definitely required or not, so tests are failing again, but hopefully trivially).

Help still needed.

**Comment by ThomasG77** [comment]
Can you point me to a generated sample online so I can find/debug where issue happened to help you?

**Comment by tomchadwin** [comment]
I'll upload one tomorrow. Or you can recreate: open one of the layers in the plugin `/test` folder, start the plugin, select a field in the `Layer search` dropdown, and export.

**Comment by tomchadwin** [comment]
No, it's not the polyfills, as I'm previewing in Chrome, which I don't think needs them - the polyfills.io request comes back with no functions.

**Comment by tomchadwin** [comment]
Here's an export. Any ideas what's wrong?

[ol3layerSearch.zip](https://github.com/tomchadwin/qgis2web/files/469315/ol3layerSearch.zip)

**Comment by ThomasG77** [comment]
Yes, if source already with a state to ready, the line https://github.com/webgeodatavore/ol3-search-layer/blob/master/src/js/index.js#L89 never fired. Unfortunately, it's not consistent. Between your usage and my demo, the quick fix for one case breaks the other...

**Comment by tomchadwin** [comment]
What's the fix for my case? If it's consistent in qgis2web exports, that will work.

**Comment by ThomasG77** [comment]
I've done a new version. Get the update on my repo.

You will also need to generate an id in the data source.
I would recommand a serial because you can't guess GeoJSON id depending of the people data. This second requirements is for centering because I use the feature id to make things work (http://openlayers.org/en/master/apidoc/ol.source.Vector.html#getFeatureById)

Below source should become


```
var geojson_lakes = {
"type": "FeatureCollection",
"crs": { "type": "name", "properties": { "name": "urn:ogc:def:crs:OGC:1.3:CRS84" } },
"features": [
{ "type": "Feature", "properties": { "cat": "1", "NAMES": "Teshekpuk Lake", "AREA_MI": 325.633, "xlabel": 28399.0, "ylabel": 7529752.0, "rotation": null }, "geometry": { "type": "Polygon", "coordinates": [...] } },
...
]
}
```
should become
```
var geojson_lakes = {
"type": "FeatureCollection",
"crs": { "type": "name", "properties": { "name": "urn:ogc:def:crs:OGC:1.3:CRS84" } },
"features": [
{ "type": "Feature", "id": 1, "properties": { "cat": "1", "NAMES": "Teshekpuk Lake", "AREA_MI": 325.633, "xlabel": 28399.0, "ylabel": 7529752.0, "rotation": null }, "geometry": { "type": "Polygon", "coordinates": [...] } },
...
]
}
```
You can note the newly added `"id": 1`.

**Comment by tomchadwin** [comment]
Hrm. I don't want to change the user data if at all possible. I guess I could namespace an `id` column if I had to - `qgis2web_id`, or something. But I'd really rather not. I'll look into it. Thanks.

**Comment by ThomasG77** [comment]
Change done.
Everything should work fine without further change than taking again the latest code.

**Comment by tomchadwin** [comment]
Ah, brilliant. Thank you! Now works in Chrome. Some issues (using the airports QGIS test dataset):

1. does not zoom to clicked result - only highlights
2. doesn't work in qgis2web preview window (QWebkit)
3. highlight is only removed when another result is clicked - should go when click anywhere else

I'd also like to display the feature's popup (if it has one), but I guess that's an implementation issue.

I can live with 2 (though obviously if it could be fixed, that would be great).

Very grateful for all of this work - obviously happy to switch discussion over to your repo. Thanks again!

**Comment by tomchadwin** [comment]
A question for all: is the GUI label `Layer search` a good one? Possible alternatives:

- feature search
- field search
- attribute search
- column search

I think `Layer search` is probably good, but thought I'd check.

**Comment by tomchadwin** [comment]
I also think we're just about ready for a release with this now.

**Comment by boesiii** [comment]
Yes I like layer search

**Comment by tomchadwin** [comment]
Added to master - please test

**Comment by tomchadwin** [comment]
Released as 1.15.0 and pushed to QGIS plugins repo.
